src/transformers/models/vits/modeling_vits.py

```diff
@@ -0,0 +1,1413 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" PyTorch VITS model."""
+
+import math
+from dataclasses import dataclass
+from typing import Optional, Tuple, Union
+
+import numpy as np
+import torch
+import torch.utils.checkpoint
+from torch import nn
+
+from ...activations import ACT2FN
+from ...deepspeed import is_deepspeed_zero3_enabled
+from ...modeling_outputs import (
+    BaseModelOutput,
+    ModelOutput,
+)
+from ...modeling_utils import PreTrainedModel
+from ...utils import add_start_docstrings, add_start_docstrings_to_model_forward, logging, replace_return_docstrings
+from .configuration_vits import VitsConfig
+
+
+logger = logging.get_logger(__name__)
+
+
+# General docstring
+_CONFIG_FOR_DOC = "VitsConfig"
+
+
+VITS_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "Matthijs/mms-tts-eng",
+    # See all VITS models at https://huggingface.co/models?filter=vits
+]
+
+
+@dataclass
+class VitsModelOutput(ModelOutput):
+    """
+    Describes the outputs for the VITS model.
+
+    Args:
+        audio (`torch.FloatTensor` of shape `(batch_size, 1, sequence_length)`):
+            Audio waveform predicted by the model.
+        sequence_lengths  (`torch.FloatTensor` of shape `(batch_size,)`):
+            The length in samples of each element in the `audio` batch.
+    """
+
+    audio: torch.FloatTensor = None
+    sequence_lengths: torch.FloatTensor = None
+
+
+# Copied from transformers.models.bart.modeling_bart._expand_mask
+def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int] = None):
+    """
+    Expands attention_mask from `[bsz, seq_len]` to `[bsz, 1, tgt_seq_len, src_seq_len]`.
+    """
+    bsz, src_len = mask.size()
+    tgt_len = tgt_len if tgt_len is not None else src_len
+
+    expanded_mask = mask[:, None, None, :].expand(bsz, 1, tgt_len, src_len).to(dtype)
+
+    inverted_mask = 1.0 - expanded_mask
+
+    return inverted_mask.masked_fill(inverted_mask.to(torch.bool), torch.finfo(dtype).min)
+
+
+@torch.jit.script
+def fused_add_tanh_sigmoid_multiply(input_a, input_b, num_channels):
+    num_channels_int = num_channels[0]
+    in_act = input_a + input_b
+    t_act = torch.tanh(in_act[:, :num_channels_int, :])
+    s_act = torch.sigmoid(in_act[:, num_channels_int:, :])
+    acts = t_act * s_act
+    return acts
+
+
+class VitsPosteriorEncoder(nn.Module):
+    def __init__(self, config: VitsConfig):
+        super().__init__()
+        self.out_channels = config.flow_size
+
+        self.conv_pre = nn.Conv1d(config.spectrogram_bins, config.hidden_size, 1)
+        self.wavenet = VitsWaveNet(config, num_layers=16)
```

Comment:
In the original VITS code that number of layers is fixed indeed.

Reply:
Seems arbitrary. What about all the other places in the code that use hardcoded channel counts, kernel sizes, etc — do those all need to go into the config too?

Reply:
The model config has to strike a balance between being able to configure the key parts of the model, but not having every single parameter such that the number of config attributes explodes.

Here, a user is almost entirely unlikely to ever want to change the channel counts / kernel sizes / etc since they'd have to completely change the structure of the weights by changing the dimensionality of all layers in the model so that they're compatible.

Whereas changing the number of layers would be a reasonable modification that is not unlikely to happen (no change to the weight structure, just expand / reduce the number of layers to match your compute / training data)